packages/core/src/dialects/abstract/query-generator-typescript.ts

```diff
@@ -126,6 +133,184 @@ export class AbstractQueryGeneratorTypeScript {
+        constraintSnippet = `CONSTRAINT ${constraintName} PRIMARY KEY (${fieldsSqlQuotedString})`;
+        if (options.deferrable) {
+          const deferOptions = { ...options, deferrable: options.deferrable };
+          constraintSnippet += ` ${this.deferConstraintsQuery(deferOptions)}`;
+        }
+
+        break;
+      case 'FOREIGN KEY': {
+        if (!this.dialect.supports.constraints.foreignKey) {
+          throw new Error(`Foreign key constraints are not supported by ${this.dialect.name} dialect`);
+        }
+
+        const references = options.references;
+        if (!references || !references.table || !(references.field || references.fields)) {
+          throw new Error('Invalid foreign key constraint options. `references` object with `table` and `field` must be specified');
+        }
+
+        const referencedTable = this.extractTableDetails(references.table);
+        constraintName = this.quoteIdentifier(options.name || `${table.tableName}_${fieldsSqlString}_${referencedTable.tableName}_fk`);
+        const quotedReferences
+          = typeof references.field !== 'undefined'
+          ? this.quoteIdentifier(references.field)
+          : references.fields!.map(f => this.quoteIdentifier(f)).join(', ');
+        const referencesSnippet = `${this.quoteTable(references.table)} (${quotedReferences})`;
+        constraintSnippet = `CONSTRAINT ${constraintName} `;
+        constraintSnippet += `FOREIGN KEY (${fieldsSqlQuotedString}) REFERENCES ${referencesSnippet}`;
+        if (options.onUpdate) {
+          if (!this.dialect.supports.constraints.onUpdate) {
+            throw new Error(`Foreign key constraint with onUpdate is not supported by ${this.dialect.name} dialect`);
+          }
+
+          constraintSnippet += ` ON UPDATE ${options.onUpdate.toUpperCase()}`;
+        }
+
+        if (options.onDelete) {
+          constraintSnippet += ` ON DELETE ${options.onDelete.toUpperCase()}`;
+        }
+
+        if (options.deferrable) {
+          const deferOptions = { ...options, deferrable: options.deferrable };
+          constraintSnippet += ` ${this.deferConstraintsQuery(deferOptions)}`;
+        }
+
+        break;
+      }
+
+      default: throw new Error(`Constraint type ${options.type} is not supported by ${this.dialect.name} dialect`);
+    }
+
+    return constraintSnippet;
+  }
+
+  removeConstraintQuery(
+    _tableName: TableNameOrModel,
+    _constraintName: string,
+    _options?: RemoveConstraintQueryOptions,
+  ): string {
+    throw new Error(`removeConstraintQuery has not been implemented in ${this.dialect.name}.`);
+  }
+
+  setConstraintQuery(columns: string[], type: 'DEFERRED' | 'IMMEDIATE') {
+    let columnFragment = 'ALL';
+
+    if (columns) {
+      columnFragment = columns.map(column => this.quoteIdentifier(column)).join(', ');
+    }
+
+    return `SET CONSTRAINTS ${columnFragment} ${type}`;
+  }
+
+  setDeferredQuery(columns: string[]) {
```

Comment:
Do we need this function?
Can we replace with a call to `setConstraintQuery`?

Reply:
We can drop it and replace it with `setConstraintQuery`, but then I propose to rename it to `setConstraintCheckingQuery`

Reply:
Do we actually need `setConstraintQuery` in the `AbstractQueryGenerator` or can this be moved to `Deferrable` as this is the only class that is using the method.

Reply:
I think it needs to stay in the query generator because dialects can't replace the implementation of the Deferrable classes otherwise

But it's still weird because:

- `setConstraintCheckingQuery` & `deferConstraintsQuery` do pretty much the same thing
- `deferConstraintsQuery` serves an ugly dual purpose where it can generate both the SQL for defining a column, and the SQL for setting the constraint checking of the current connection. This is because the `Deferrable` namespace is a mix of both when really `INITIALLY_DEFERRED`, `INITIALLY_IMMEDIATE` and `NOT` should be their own enum, while `SET_DEFERRED` and `SET_IMMEDIATE` should be classes that belong to a namespace called something like `ConstraintCheckingBehavior` (ok that's maybe too verbose)

Then we could have the following:

- Delete `deferConstraintsQuery`
- Overload `setConstraintCheckingQuery` to make it accept two signatures:
  - signature 1: a single instance of `ConstraintCheckingBehavior`
  - signature 2: a `ConstraintCheckingBehavior` class + optional column array
  - It would not call `.toSql` on the target class anymore, but generate the SQL itself
- Add `_getDeferrableConstraintSnippet` (private/protected depending on whether sub-dialects need access to it) which accepts the `Deferrable` enum and generate the correct sql based on it.

What do you think?

Reply:
I've tested these changes however I have some questions.

 - Do you want to remove the `Deferrable` class and replace it with the below code
 - For `setConstraintCheckingQuery` should this just call `_getDeferrableConstraintSnippet` when the deferrable parameter is `ConstraintChecking` enum which would return the generated SQL.


```ts
import isEqual from 'lodash/isEqual';
import { classToInvokable } from './utils/class-to-invokable.js';
import { EMPTY_ARRAY } from './utils/object.js';

/**
 * Can be used to make foreign key constraints deferrable.
 * This is only supported in PostgreSQL.
 *
 * The foreign keys can be configured like this. It will create a foreign key
 * that will check the constraints immediately when the data was inserted.
 *
 * ```js
 * class MyModel extends Model {}
 * MyModel.init({
 *   foreign_id: {
 *     type: DataTypes.INTEGER,
 *     references: {
 *       model: OtherModel,
 *       key: 'id',
 *       deferrable: Sequelize.ConstraintChecking.INITIALLY_IMMEDIATE
 *     }
 *   }
 * }, { sequelize });
 * ```
 */
export enum ConstraintChecking {
  INITIALLY_DEFERRED = 'INITIALLY_DEFERRED',
  INITIALLY_IMMEDIATE = 'INITIALLY_IMMEDIATE',
  NOT = 'NOT',
}

/**
 * Can be used to set constraints deferrable within a transaction.
 * This is only supported in PostgreSQL.
 *
 * The constraints can be configured to be deferrable in a transaction like this.
 * It will trigger a query once the transaction has been started and set the constraints
 * to be checked at the very end of the transaction.
 *
 * ```js
 * sequelize.transaction({
 *   deferrable: Sequelize.ConstraintCheckingBehavior.SET_DEFERRED
 * });
 * ```
 */
export class ConstraintCheckingBehavior {
  isEqual(_other: unknown): boolean {
    throw new Error('isEqual implementation missing');
  }

  toString() {
    throw new Error('toString implementation missing');
  }

  /**
   * Will trigger an additional query at the beginning of a
   * transaction which sets the constraints to deferred.
   */
  static readonly SET_DEFERRED = classToInvokable(class SET_DEFERRED extends ConstraintCheckingBehavior {
    readonly #constraints: readonly string[];
    readonly #name = 'DEFERRED';

    /**
     * @param constraints An array of constraint names. Will defer all constraints by default.
     */
    constructor(constraints: readonly string[] = EMPTY_ARRAY) {
      super();
      this.#constraints = Object.freeze([...constraints]);
    }

    isEqual(other: unknown): boolean {
      return other instanceof SET_DEFERRED && isEqual(this.#constraints, other.#constraints);
    }

    toString() {
      return this.#name;
    }
  });

  /**
   * Will trigger an additional query at the beginning of a
   * transaction which sets the constraints to immediately.
   */
  static readonly SET_IMMEDIATE = classToInvokable(class SET_IMMEDIATE extends ConstraintCheckingBehavior {
    readonly #constraints: readonly string[];
    readonly #name = 'IMMEDIATE';

    /**
     * @param constraints An array of constraint names. Will defer all constraints by default.
     */
    constructor(constraints: readonly string[] = EMPTY_ARRAY) {
      super();
      this.#constraints = Object.freeze([...constraints]);
    }

    isEqual(other: unknown): boolean {
      return other instanceof SET_IMMEDIATE && isEqual(this.#constraints, other.#constraints);
    }

    toString() {
      return this.#name;
    }
  });
}

```

Reply:
Almost but I noticed I used inconsistent names in my comment, which may have thrown you off

- I think `ConstraintChecking` could continue to be named `Deferrable` even if it's an enum, because that's the value we use on the `deferrable` option
- `ConstraintCheckingBehavior` could be named `ConstraintChecking`
- I don't think the `#name` property is necessary, it's only ever returned by toString. Its implementation could be `return this.constructor.name`

But that's about it, so yes pretty much what you wrote :)

Reply:
Changes added now, LMK if you have any suggestions/comments.

Reply:
> would need to be DEFERRED and IMMEDIATE instead of SET_DEFERRED and SET_IMMEDIATE.

Is it because of a failing test or is there another reason? :o

Reply:
No, it means that I don't need to use a switch/if statement in the `setConstraintCheckingQuery` method to generate the SQL.

By changing the options I can just use the `toString` method for the class to populate the type in the SQL.

Reply:
That works for me